samples/examples/FSharpExamples/Matrix.fs

```diff
@@ -28,38 +28,53 @@ open Prajna.Api.FSharp
 
 open Prajna.Examples.Common
 
-type MatrixSample() =
+module Matrix =
 
     let (.*) (xs: float[]) (ys: float[]) = 
-        // simulate doing some more work, 
-        // e.g.: run the remaining layers of a neural net
-        Threading.Thread.Sleep(3) 
         Array.fold2 (fun sum x y -> sum + x * y) 0.0 xs ys 
 
     let printTiming (msg: string) (x: Lazy<'b>) : 'b =
         let sw = Stopwatch.StartNew()
+        printf "%s: " msg
         let ret = x.Value
         sw.Stop()
-        printfn "%s: %A" msg sw.Elapsed
+        printfn "%A" sw.Elapsed
         ret
 
-    let run (cluster: Cluster) =
+    let random = new Random(0)
+
+    let randomVector n =  Array.init n (fun _ -> random.NextDouble())
+    
+    let randomMatrix m n =  Array.init m (fun _ -> randomVector n)
+
+    /// Matrix-Vector multiplication for illustrational purposes.
+    /// The overhead of distributing makes it faster to run this locally.
+    /// See Matrix-Matrix for a case where the computation is actually faster distributed.
+    let matrixVector (cluster: Cluster) =
 
         let numRows = 200
         let vectorSize = 100
 
-        let r = new Random(0)
-        let randomMatrix =  Array.init numRows (fun _ -> Array.init vectorSize (fun _ -> r.NextDouble()))
-        let vector = Array.init vectorSize (fun _ -> r.NextDouble())
+        let matrix =  randomMatrix numRows vectorSize
+        let vector = randomVector vectorSize
 
         let singleCoreResult = 
             printTiming "single core" <| lazy 
-                randomMatrix |> Array.map (fun row -> row .* vector)
-        
+                matrix |> Array.map (fun row -> row .* vector)
+
+        let numPartitions = Environment.ProcessorCount
+        let dSet =
+            printTiming "distributing" <| lazy
+                let d =
+                    DSet<float[]>(Cluster = cluster, NumParallelExecution = numPartitions)
+                    |> DSet.distributeN numPartitions matrix
```

Comment:
I guess I didn't make myself clear.  First, the "fold" on "d" will not cause "d" to be persisted to memory. When you use "d" in other places, the operations that computes "d" (e.g. distributeN) will be re-evaluated. So you don't achieve what you want.  Secondly, "cacheInMemory" itself is not an action thus it will not be evaluated unless there is an action following it.  That's why itself alone does not help either. 

What you need is


        let dSet =
            printTiming "distributing" <| lazy
                let d =
                    DSet<float[]>(Cluster = cluster, NumParallelExecution = numPartitions)
                    |> DSet.distributeN numPartitions matrix
                    |> DSet.cacheInMemory
                // Force push DSet to cluster
                do d |> DSet.fold (fun _ _ -> 0) (fun _ _ -> 0) 0 |> ignore
                d

That is, you use DSet.fold to trigger it to be cached. Thus the returned "d" is the cached one. When it was used in other computation graph, it will not need to be re-evaluated from the root. 

This is how it is supposed to work now. If you see anything unexpected, then there're something wrong.  But this behavior was verified in UT.  See https://github.com/BrunoBozza/Prajna/blob/master/tests/CoreLib/DSet.fs#L1289